components/support/rc_crypto/nss/src/pbkdf2.rs

```diff
@@ -0,0 +1,131 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use crate::error::*;
+use crate::util::ensure_nss_initialized;
+use nss_sys::SECOidTag;
+
+// ***** BASED ON THE FOLLOWING IMPLEMENTATION *****
+// https://github.com/sqlcipher/sqlcipher/blob/4a81bea61e1da6fec222d713852830f1fd01aed2/src/crypto_nss.c#L182-L232
+// Which I believe is based on how it is done on WebCrypto:
+// https://searchfox.org/mozilla-central/rev/8ccea36c4fb09412609fb738c722830d7098602b/dom/crypto/WebCryptoTask.cpp#2567
+
+pub fn nss_pbkdf2_generate(
+    password: &mut [u8],
+    salt: &mut [u8],
+    iterations: u32,
+    out: &mut [u8],
+) -> Result<()> {
+    ensure_nss_initialized();
+    // Todo: Maybe add support for different hashing algorithms?
+    let oid_tag = SECOidTag::SEC_OID_HMAC_SHA256;
+    let mut sec_salt = nss_sys::SECItem {
+        len: salt.len() as u32,
+        data: salt.as_mut_ptr(),
+        type_: 0,
+    };
+
+    let alg_id = unsafe {
+        nss_sys::PK11_CreatePBEV2AlgorithmID(
```

Comment:
You should wrap this call in a [scoped pointer wrapper](https://searchfox.org/mozilla-central/rev/7cadba1d8b8feaec4615f5bb98aac4b8a719793c/security/manager/ssl/ScopedNSSTypes.h#304-305). We do this using the `scoped_ptr!` macro, there's some examples in components/support/rc_crypto/nss/src/pk11/types.rs.

Reply:
I had to add to the scoped_ptr! macro a second expression where the destructor has an argument (since the algorithm ID destructor requires an extra argument) wasn't sure if that is best, or if I should define a separate struct for AlgorithmID without using the macro.

Reply:
`pk11_destroy_context_true` is an example where we send a param to the destructor

Reply:
I can't believe I didn't see that 🤦 Thanks @eoger and I really appreciate your reviews 🙏